### PR TITLE
Update Metrics and Flags for GPU

### PR DESCRIPTION
**Pull request by ashors1**
This PR includes the following changes:
- Adds flags used for GPU training and GPU scripts (multiprocess flags, option to overwrite TFDS data dir, option to log compile times)
- Adds some additional logging (log compile times, log e2e time, add seqs_per_sec metric)
- Fixes logging for GA such that reported loss and throughput reflects that of a full batch rather than a subbatch 

Including all of the above changes in a single PR to avoid merge conflicts from multiple PRs

### CODE REVIEW

File: paxml/summary_utils.py [hunk 1]
```diff
@@ -396,6 +398,12 @@ def write_summary_entry(summary_writer: SummaryWriter,
       status_msg += f', steps/sec {steps_per_sec}'
       write_summary_tensor(step_i, 'Steps/sec', steps_per_sec,
                            SummaryType.AGGREGATE_SCALAR)
+
+      if global_bs is not None:
+        seqs_per_sec = num_sub_batches*global_bs*steps_per_sec
```
Comment (zhangqiaorjc) [inline paxml/summary_utils.py:403]: why is this seqs_per_sec, we are not multiplying by seq_len

File: paxml/train.py [hunk 2]
```diff
@@ -45,6 +45,7 @@
 from praxis import base_layer
 from praxis import py_utils
 from praxis import pytypes
+from statistics import mean
```
Comment (zhangqiaorjc) [inline paxml/train.py:48]: lets use numpy or jax.numpy for mean? we don't typically use python statistics module

File: paxml/train.py [hunk 3]
```diff
@@ -1114,40 +1120,48 @@ def _train_and_evaluate_common(
...
+          with py_utils.timeit() as train_period:
+            program_output = train_program.run_step(
+                partitioned_train_state,
+                train_prng_seed,
+                model_inputs,
+                train_program.train_unpadded_global_batch_size,
+            )
+            # TODO(hthu): Pass program output around instead of unpacking.
+            partitioned_train_state = program_output.state
+            loss = program_output.aux['loss']
```
Comment (zhangqiaorjc) [inline paxml/train.py:1153]: `loss` is asynchronously computed by JAX backend, see
https://jax.readthedocs.io/en/latest/async_dispatch.html#asynchronous-dispatch

when you ask for `loss` actual value, you will be blocking the main thread from dispatching the next GPU computation, so ideally the mean_loss_over_subbatches need to be async computed as well

the way to do that is avoid host computation (numpy or statistics will compute on CPU)

instead of loss_tracker=[], i recommend 

```
summed_loss = jnp.zeros((), dtype=jnp.float32)
for grad_accum_step in range(num_sub_batches):
  program_output = pjitted_train_step(...)
  loss = program_output.aux['loss']
  summed_loss = summed_loss + loss
loss = summed_loss / num_sub_batches
```

The add and divide will now happen on GPU as async dispatched computation

**Review by zhangqiaorjc — CHANGES_REQUESTED** [review]

File: paxml/train.py [hunk 4]
```diff
@@ -1114,40 +1120,48 @@ def _train_and_evaluate_common(
...
-              train_program.train_unpadded_global_batch_size,
-          )
-          # TODO(hthu): Pass program output around instead of unpacking.
-          partitioned_train_state = program_output.state
-          loss = program_output.aux['loss']
-          weighted_scalars = program_output.aux['weighted_scalars']
-          per_example_out = program_output.aux['per_example_out']
-          summary_tensors = program_output.aux['summary_tensors']
+      loss_tracker = []
+      for ga_step in range(num_sub_batches):
```
Comment (zhangqiaorjc) [inline paxml/train.py:1124]: nit: let's don't use `ga`, let's spell it out `grad_accum` everywhere

**Review by zhangqiaorjc — CHANGES_REQUESTED** [review]

File: paxml/train.py (hunk 3 above)
Comment (zhangqiaorjc) [reply paxml/train.py:1153]: ignore the last comment, `summed_loss = summed_loss + loss` should just work

there's no communications involved, so pjit=jit

File: paxml/summary_utils.py (hunk 1 above)
Comment (zhangqiaorjc) [reply paxml/summary_utils.py:403]: nvm, we don't want to multiply by seq_len, otherwise it'd be tokens/sec. This is resolved.

**Review by zhangqiaorjc — APPROVED** [review]

**Comment by zhangqiaorjc** [comment]
@ashors1 can you split train.py changes into another PR? just to make this PR smaller and easier to merge

File: paxml/main.py [hunk 5]
```diff
@@ -360,13 +380,22 @@ def run(experiment_config: base_experiment.BaseExperiment,
 
 
 def main(argv: Sequence[str]) -> None:
+  start = time.time()
   if len(argv) > 1:
     raise app.UsageError('Too many command-line arguments.')
 
+  if FLAGS.tfds_data_dir is not None:
+    seqio.set_tfds_data_dir_override(FLAGS.tfds_data_dir)
```
Comment (zhangqiaorjc) [inline paxml/main.py:388]: @ashors1 can you move the seqio here

  if FLAGS.tfds_data_dir is not None:
    # seqio import is slow so avoid module level 
    import seqio
    seqio.set_tfds_data_dir_override(FLAGS.tfds_data_dir)

**Review by zhangqiaorjc — APPROVED** [review]

File: paxml/setup_jax.py [hunk 6]
```diff
@@ -15,20 +15,32 @@
...
 from typing import Optional
 
 from absl import logging
 import jax
 from praxis import py_utils
 import tensorflow.compat.v2 as tf
 
+@dataclasses.dataclass
+class JaxDistributedOptions:
+  coordinator_address: int
```
Comment (zhangqiaorjc) [inline paxml/setup_jax.py:28]: address should be a string


def initialize(coordinator_address: Optional[str] = None,
               num_processes: Optional[int] = None,
               process_id: Optional[int] = None,
               local_device_ids: Optional[Union[int, Sequence[int]]] = None):
